doc/manpages/ndnputmeta.1
'\" t
.\"     Title: ndnputmeta
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.76.0 <http://docbook.sf.net/>
.\"      Date: 05/16/2013
.\"    Manual: \ \&
.\"    Source: \ \& 0.7.2
.\"  Language: English
.\"
.TH "NDNPUTMETA" "1" "05/16/2013" "\ \& 0\&.7\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ndnputmeta \- Associate a file with NDNx Content as metadata
.SH "SYNOPSIS"
.sp
\fBndnputmeta\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-v] [\-raw] [\-unversioned] [\-local | \-allownonlocal] [\-timeout \fImillis\fR] [\-log \fILEVEL\fR] \fIndnxname\fR \fImetaname\fR \fIfilename|url\fR
.SH "DESCRIPTION"
.sp
The \fBndnputmeta\fR utility publishes a local file \fIfilename\fR or url \fIurl\fR as metadata associated with the content \fIndnxname\fR\&. The metadata is published as as of NDNx Data in accordance with the naming conventions for segmented streams or files, optionally unversioned\&. For the default case of versioned content, \fBndnputmeta\fR will publish the metadata with the version based on the local machine time and will associate it with the latest version of the referenced NDNx content\&.
.sp
The \fIndnxname\fR must be specified using the NDNx URI encoding syntax\&. For simple cases of ASCII name components this is just pathname syntax with / delimiters\&.
.sp
The \fImetaname\fR must be the relative path of the metadata for the specified ndnxname, for example \fI\fIthumbnail\fR\fR\fI/image\fR\&.
.sp
The \fIfilename\fR must be the pathname of a local file that will published under the \fIndnxname\fR\&.
.sp
The \fIurl\fR must be a valid url to be published under a \fIndnxname\fR\&.
.sp
\fBndnputmeta\fR will publish versioned and segmented meta data objects, and the key used to sign them, to a repository by default\&. If there is not a reachable repository, \fBndnputmeta\fR will exit with a \fIwait for put drain\fR error\&. If there is a corresponding application able to retrieve the content \fBndnputmeta\fR can be used to publish the content with \fI\-raw\fR\&. However no such application is currently supported\&. \fI\-raw\fR allows \fBndnputmeta\fR to skip the repository handshake and directly put the content objects to satisfy the interests expressed by the reader application\&.
.sp
\fBndnputmeta\fR can also be used to publish metadata with a specific key using \fI\-as\fR where \fIpathToKeystore\fR must be a valid path to the local filesystem where a user key has previously been stored\&.
.sp
\fBndnputmeta\fR can additionally be used with access control using the \fI\-ac\fR flag\&.
.SH "OPTIONS"
.PP
\fB\-v\fR
.RS 4
Run
\fBndnputmeta\fR
in verbose mode\&.
.RE
.PP
\fB\-raw\fR
.RS 4

\fBndnputmeta\fR
will publish the metadata without contacting a repository\&. This mode will fail unless there is another application, capable of reading the metadata, running\&.
.RE
.PP
\fB\-unversioned\fR
.RS 4
Do not append a version component to the supplied
\fImetaname\fR\&. This option may be used when the supplied
\fImetaname\fR
already includes an explicit version component at the end (which must be URI encoded)\&. This option may also be used for metadata content that is organized according to the segmentation convention but does not use the versioning convention at all\&. When this option is used, the FileOutputStream abstraction will not be used, since that implies versioning\&.
.RE
.PP
\fB\-local\fR
.RS 4
Specifies that a local repository be used to save the file\&. This is the (new, as of release 0\&.5\&.0) default behavior\&.
.RE
.PP
\fB\-allownonlocal\fR
.RS 4
Specifies that either a local or non\-local repository be used to save the file\&.
.RE
.PP
\fB\-timeout\fR \fImillis\fR
.RS 4
Set the timeout that will be used for putting content objects on during the stream reading\&. I\&.e\&., the time to wait for incoming interests\&.
.RE
.PP
\fB\-log\fR \fILEVEL\fR
.RS 4
Set the logging level\&.
\fILEVEL\fR
must be the an acceptable Java Logger level: either an integer or a level name (e\&.g\&. SEVERE)\&.
.RE
.PP
\fB\-debug\fR \fIportno\fR
.RS 4
Allows attachment of an eclipse remote debugger to the utility at the specified port number\&.
.RE
.PP
\fB\-javaopts\fR \fIoptions\fR
.RS 4
Allows use of arbitrary java properties or options with the utility\&.
.RE
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure (syntax or usage error; publishing error \- no processes running that are prepared to ask or already asking for content)
.RE
.SH "EXAMPLE"
.sp
The following example publishes a new version of the file "mydocument" and adds the file "image\&.png" as thumbnail metadata for it
.sp
.if n \{\
.RS 4
.\}
.nf
ndnputfile "ndn:/mydocument" "mydocument"
ndnputmeta "ndn:/mydocument" "_thumbnail_/image\&.png" "image\&.png"
.fi
.if n \{\
.RE
.\}
.SH "AUTHOR"
.sp
Paul Rasmussen


CHANGELOG
NDNx v0.1 (2013-08-23)
--------

Initial public release.

Based on CCNx version 0.7.2. 
https://github.com/ProjectCCNx/ccnx/tree/ccnx-0.7.2


javasrc/src/main/org/ndnx/ndn/utils/ndnputmeta.java
/*
 * A NDNx command line utility.
 *
 * Portions Copyright (C) 2013 Regents of the University of California.
 * 
 * Based on the CCNx C Library by PARC.
 * Copyright (C) 2008, 2009, 2011, 2012 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

package org.ndnx.ndn.utils;

import java.io.IOException;
import java.security.InvalidKeyException;
import java.util.logging.Level;

import org.ndnx.ndn.NDNHandle;
import org.ndnx.ndn.config.ConfigurationException;
import org.ndnx.ndn.impl.support.Log;
import org.ndnx.ndn.profiles.VersioningProfile;
import org.ndnx.ndn.profiles.metadata.MetadataProfile;
import org.ndnx.ndn.protocol.ContentName;
import org.ndnx.ndn.protocol.MalformedContentNameStringException;

/**
 * Command-line utility to write metadata associated with an existing file in ndnd. The "metaname" should
 * be the relative path (including filename) for the desired metadata only.
 * By default this writes to the repo. Otherwise there must be a corresponding ndngetfile to retrieve
 * the data.
 **/
 public class ndnputmeta extends CommonOutput implements Usage {
	 static ndnputmeta ndnputmeta = new ndnputmeta();

	/**
	 * @param args
	 */
	public void write(String[] args) {
		Log.setDefaultLevel(Level.WARNING);
		for (int i = 0; i < args.length; i++) {
			if (CommonArguments.parseArguments(args, i, ndnputmeta)) {
				i = CommonParameters.startArg;
				continue;
			}
			if ((i + 3) >= args.length) {
				CommonParameters.startArg = i;
				break;
			}
			if (args[i].equals("-local")) {
				CommonParameters.local = true;
			} else if (args[i].equals(("-allownonlocal"))) {
				CommonParameters.local = false;
			} else if (args[i].equals(("-raw"))) {
				CommonParameters.rawMode = true;
			} else
				usage(CommonArguments.getExtraUsage());
		}

		if (args.length != CommonParameters.startArg + 3) {
			usage(CommonArguments.getExtraUsage());
		}

		long starttime = System.currentTimeMillis();
		try {
			// If we get one file name, put as the specific name given.
			// If we get more than one, put underneath the first as parent.
			// Ideally want to use newVersion to get latest version. Start
			// with random version.

			ContentName baseName = ContentName.fromURI(args[CommonParameters.startArg]);
			String metaArg = args[CommonParameters.startArg + 1];
			if (!metaArg.startsWith("/"))
				metaArg = "/" + metaArg;
			ContentName metaPath = ContentName.fromURI(metaArg);
			NDNHandle handle = NDNHandle.open();
			ContentName prevFileName = MetadataProfile.getLatestVersion(baseName, metaPath, CommonParameters.timeout, handle);
			if (null == prevFileName) {
				System.out.println("File " + baseName + " does not exist");
				System.exit(1);
			}
			ContentName fileName = VersioningProfile.updateVersion(prevFileName);
			if (CommonParameters.verbose)
				Log.info("ndnputmeta: putting metadata file " + args[CommonParameters.startArg + 1]);

			doPut(handle, args[CommonParameters.startArg + 2], fileName);
			System.out.println("Inserted metadata file: " + args[CommonParameters.startArg + 1] + " for file: " + args[CommonParameters.startArg] + ".");
			if (CommonParameters.verbose)
				System.out.println("ndnputmeta took: "+(System.currentTimeMillis() - starttime)+" ms");
			System.exit(0);
		} catch (ConfigurationException e) {
			System.out.println("Configuration exception in put: " + e.getMessage());
			e.printStackTrace();
		} catch (MalformedContentNameStringException e) {
			System.out.println("Malformed name: " + args[CommonParameters.startArg] + " " + e.getMessage());
			e.printStackTrace();
		} catch (IOException e) {
			System.out.println("Cannot put metadata file. " + e.getMessage());
			e.printStackTrace();
		} catch (InvalidKeyException e) {
			System.out.println("Cannot publish invalid key: " + e.getMessage());
			e.printStackTrace();
		}
		System.exit(1);

	}

	@Override
	public void usage(String extraUsage) {
		System.out.println("usage: ndnputmeta " + extraUsage + "[-v (verbose)] [-raw] [-unversioned] [-local | -allownonlocal] [-timeout millis] [-log level] [-as pathToKeystore] [-ac (access control)] <ndnname> <metaname> (<filename>|<url>)*");
		System.exit(1);
	}

	public static void main(String[] args) {
		new ndnputmeta().write(args);
	}
}
